Add support for Linkerd 2.13

alpeb opened the pull request:
In Linkerd 2.13 the Prometheus instance in the `linkerd-viz` namespace is now locked behind an
[_AuthorizationPolicy_](https://github.com/linkerd/linkerd2/blob/stable-2.13.1/viz/charts/linkerd-viz/templates/prometheus-policy.yaml) that only allows access to the `metrics-api` _ServiceAccount_.

This adds an extra _AuthorizationPolicy_ to authorize the `flagger` _ServiceAccount_. This also implies that the Flagger workload has to be injected by the Linkerd proxy, and that can't happen in the same `linkerd` namespace where the control plane lives, so we're moving Flagger into the new injected `linkerd-flagger` namespace.

The `namespace` field in `kustomization.yml` was resetting the namespace for the new _AuthorizationPolicy_ resource, so that gets restored back to `linkerd-viz` using a `patchesJson6902` entry. A better way to do this would have been to use the `unsetOnly` field in a _NamespaceTransformer_ (see kubernetes-sigs/kustomize#4708) but for the life of me I couldn't make that work...

aryan9600 commented:
hey @alpeb, thanks for opening this PR! could you also bump the linkerd version used in the e2e tests (see `test/linkerd/install.sh`)

alpeb commented on test/linkerd/install.sh:
Oh dang, sorry about that! I've just pushed another update with that :pray:

stefanprodan commented:
I would prefer to use the `flagger-system` namespace which is part of the default installation. To @aryan9600 point changing the namespace and adding a custom resource is major breaking change. Let's add the `AuthorizationPolicy` as an opt-in to Flagger Helm chart too.

alpeb commented:
Thanks for all your feedback :+1:

In my latest push I've changed the namespace to `flagger-system` and updated the Helm chart with a couple of additional entries in the `values.yaml` file.
I also reorganized the commits a little bit.

aryan9600 commented:
okay so tests are finally passing :)))
since there are so many Deployments that linkerd installs, CPU requests were getting exhausted, so I had to scale down a few unnecessary Deployments in the `linkerd-viz` ns  and delete the `podinfo` Daemonset since we don't need it for linkerd e2e anyway. @alpeb  it'd be great if you could squash these commits into 1-2 meaningful commits and then this should be ready to merge.

alpeb commented:
Thanks for having taken care of that test! :sweat_smile: 
I've updated the linkerd tutorial and squashed the commits :+1:

aryan9600 commented on docs/gitbook/tutorials/linkerd-progressive-delivery.md:
```diff
@@ -8,20 +8,53 @@ This guide shows you how to use Linkerd and Flagger to automate canary deploymen
+  --n flagger-system \
+  --set meshProvider=linkerd \
+  --set metricsServer=http://prometheus.linkerd-viz:9090 \
+  --set linkerdAuthPolicy.create=true
```
could you also add a note here explaining that this is only needed if you're on linkerd 2.12 and later?

alpeb replied:
Done :+1:

aryan9600 reviewed (approved):
lgtm! thanks @alpeb 🙇

stefanprodan reviewed (approved):
LGTM

Thanks @alpeb 

PS. @aryan9600 please make sure we document the breaking changes in the changelog, so people know they must delete the Flagger deployment from the linkerd namespace.
